### PR TITLE
StreamSystem Refactoring

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich !

On top of your https://github.com/c5t/Current/pull/386:

Done:
* `Persister` can now be iterated upon.
* No more `Clone` (except for `Variant<>`).
* No more `aloha`.
* No more `PublishDerived` and `can_be_stored_in_unique_ptr`.

Left to do:
* Proper waiting in Sherlock (it's a spin lock now to have it run).
* `// TODO(dkorolev) + TODO(mzhurovich): Test IDX_TS-es of the persisted data too.`
* ScopeOwned, for Persister and for MMQ.
* Assess feasibility of `static_assert(current::ss::IsStreamSubscriber<F, ENTRY>::value, "");` in Sherlock's `ListenerThread`.

Thanks,
Dima

### CODE REVIEW

File: Blocks/HTTP/impl/java_client.h [hunk 1]
```diff
@@ -39,7 +39,7 @@ SOFTWARE.
 
 // Wrapper over native Android/Java code for HTTP GET/POST.
 
-namespace aloha {
+namespace apple_wrapper {
```
Comment (mzhurovich) [inline Blocks/HTTP/impl/java_client.h]: s/apple_wrapper/java_wrapper/

File: Storage/persister/sherlock.h [hunk 2]
```diff
@@ -51,7 +51,7 @@ class SherlockStreamPersisterImpl<TypeList<TS...>, PERSISTER> {
   void PersistJournal(MutationJournal& journal) {
     if (!journal.commit_log.empty()) {
       T_TRANSACTION transaction;
-      for (auto&& entry : journal.commit_log) {
+      for (auto& entry : journal.commit_log) {
```
Comment (mzhurovich) [inline Storage/persister/sherlock.h]: `auto&&` should do what it's supposed to do.

File: Sherlock/sherlock.h [hunk 3]
```diff
@@ -202,15 +209,20 @@ class StreamImpl {
   template <typename F>
   class ListenerThread {
    private:
+    // static_assert(current::ss::IsStreamSubscriber<F, ENTRY>::value, "");
```
Comment (mzhurovich) [inline Sherlock/sherlock.h]: `|| (is_unique_ptr... && IsStreamSubscriber<F::element_type...)`?

**Comment by dkorolev** [comment]
@mzhurovich Fixed everything (thanks!), refactored Sherlock to be a bit cleaner, and added persister-level timestamp tests.

PTAL!

Thanks,
Dima

**Pull request merged**